### ISSUE TITLE
Killzone liberation - slow framebuffer CLUT effects, similar to Sonic

### ISSUE THREAD

**Issue by daniel229** [hrydgard/ppsspp#6207]
Update:
Tested with 1.0-1 build,issue 2 , 4 and 5 is still present.
1. fixed ~~since the framebuffer-clut merged https://github.com/hrydgard/ppsspp/pull/5767 last working build is v0.9.8-865-g585050d~~
   ~~before~~
   (https://cloud.githubusercontent.com/assets/3481559/3141008/15f10410-e95c-11e3-9b45-15dc9b750324.png)

~~after~~
(https://cloud.githubusercontent.com/assets/3481559/3141007/15ec311a-e95c-11e3-983e-22144e12721f.png)

**Comment by daniel229** [comment]
Sunday x Magazine Shuuketsu Choujou Daikessen hangs after video same as killzone,since https://github.com/hrydgard/ppsspp/pull/5245

**Comment by BinkanSalaryman** [comment]
Any solution to this after 3 years? It's impossible to play on phone, and the controls for PC are hard 🌵

**Comment by weihuoya** [comment]
there are a lot of `GE_PRIM_TRIANGLE_STRIP` and `GE_PRIM_RECTANGLES` draw for post process, and these can safely remove.

**Comment by weihuoya** [comment]
find a dirty hack for this. just skip prim that draw offscreen or copy from offscreen.
and then it run 30fps on my phone, still can't run full speed 60fps.
```c++
void GPUCommon::Execute_Prim(u32 op, u32 diff) {
	...
	void *verts = Memory::GetPointerUnchecked(gstate_c.vertexAddr);
	//if (PSP_CoreParameter().compat.flags().KillZoneHack) {
		if (prim == GE_PRIM_RECTANGLES) { // this for game scene
			short* pos = (short*)verts;
			if (pos[0] > 479 || pos[2] > 479) {
				return;
			}
		}
		else if (prim == GE_PRIM_TRIANGLE_STRIP) { // this for game menu
			if (0x12000183 == gstate.vertType) {
				return;
			}
		}
	//}
	...
}
```

**Comment by unknownbrackets** [comment]
But isn't it drawing off screen for a reason?  I remember seeing something at some point that suggested to me that it was using the area between strides to render temporary effects, and then texturing that to the screen.

The effect might not be working, and might also be why the light animation doesn't move at the correct pace.

I really hope we don't go the route of PCSX2 and add a library of "skip code" hacks to make games render faster / hack around graphics issues... that's like when we've given up.

-[Unknown]

**Comment by hrydgard** [comment]
It's a sliding scale between purity and speedhacks but I like the Sonic Rivals-style command list post-processing I did for Vulkan better than skip-hacks, at least as we try to preserve behaviour completely, just rearrange things to not be ridiculously inefficient with modern APIs. It might be possible to do one of those for this game too.

**Comment by hrydgard** [comment]
Will look at this again for 1.8.0 but might bounce it to 1.9.0 if time runs out.

**Comment by Panderner** [comment]
Here's a GE Dump for title screen
[DUMP.ppdmp.zip](https://github.com/hrydgard/ppsspp/files/5955944/DUMP.ppdmp.zip)

**Comment by Gamemulatorer** [comment]
This game is the most lag games on ppsspp?
Updated GE Dump
[killzone.ppdmp.zip](https://github.com/hrydgard/ppsspp/files/7030060/killzone.ppdmp.zip)

**Comment by Panderner** [comment]
Killzone menu screens are GPU intensive for reasons i tested from my GTX 1050 Ti:

in 1x resolution the GPU usage is average of 60% using DirectX 11:
![Screenshot (7)](https://user-images.githubusercontent.com/37503397/181718898-70ea0304-db98-4cae-b587-4e29609905bb.png)



But in 2x resolution the GPU hits 99% and the game lags.
![Screenshot (6)](https://user-images.githubusercontent.com/37503397/181718909-1cd3599b-0970-46f6-acdf-94dc3f382aca.png)

**Comment by hrydgard** [comment]
Killzone general rendering process

Title Screen
5551 framebuffer at 0 (displayed), 8888 Framebuffer at 44000, Z buffer at cc000 (unused). format 8888

standard clear of the left 480x272
rect draw of logo to left 480x272
Some simple text drawn.

Now, first clearing, then texturing from 44000 while writing to 44000, but only
writing to the stride gap (rightmost 32x272 pixels). 

The clear is a rect draw, then there are two triangle strips.
x-range 480-512,
y-range 0-135.997925  (half of 272)

Rect draw, easily found (x1 = 480, etc).

```
X	Y	Z	U	V	Color	NX	NY	NZ
480.000000	136.997803	32767.000000	480.046875	136.500000	00000000	0.000000	0.000000	1.000000
512.000000	136.997803	32767.000000	512.046875	136.500000	00000000	0.000000	0.000000	1.000000
480.000000	270.995850	32767.000000	480.046875	270.500000	00000000	0.000000	0.000000	1.000000
512.000000	270.995850	32767.000000	512.046875	270.500000	00000000	0.000000	0.000000	1.000000
```

The Y and U coordinates vary a bit (1-2 pixels). X coordinates stay the same (exactly 480, 512).

After doing this a bunch of time, texture from this margin and draw back to the main part of the framebuffer.
Scissor is used nicely, 0,0->480x272 for the main, 480,0->32x272 for the border.

In-game, rendering the scene within a 480x272 scissor, then it's doing black-and-white copies to the margin using CLUT32, but never reading them? When doing this, the scissor is 512x272.

Anyway in conclusion, I think with a compat flag and some scissor rect checks, we can fairly easily do a hack that splits the main framebuffer, which should make this way, way faster. Not sure it'll work in all parts of the game though, it might be that we have to go peek in vertex coordinates. Will give it a shot soon.

**Comment by hrydgard** [comment]
Have something that kinda works but needs a bit more tweaking.

Also noticed that this game seems to have broken timing and renders each frame multiple times, similarly to GoW making it extra slow, but our ForceMax60Fps isn't working because it never moves the display pointer, instead it blits from an 8888 framebuffer onto a 5551 one for display. Will need another tweak...

**Comment by Panderner** [comment]
But it seems the light effects are missing
![UCUS98646_00000](https://user-images.githubusercontent.com/37503397/187623362-1a745477-3632-460e-bbcb-c4899d1600b6.jpg)

**Comment by hrydgard** [comment]
I'll have a look at that, might be only on the title screen. The effect is definitely working in-game at least.

**Comment by hrydgard** [comment]
Yeah, I failed to notice that the title screen's method of drawing is actually quite different, and even uses vertices not in through-mode. Will require some additional tweaks. Hopefully there are not even more margin rendering paths later in the game :(

**Comment by hrydgard** [comment]
Title screen is going to be harder to fix, due to the use of non-through drawing. Still very doable, but I'm gonna do some other things in between I think.

**Comment by Panderner** [comment]
Unfortunately still not fixed the light effect as of v1.13.1-633-g437d6d30a

**Comment by hrydgard** [comment]
Yeah that's known, it'll be tricky to fix fully. Taking a short break from Killzone right now as we achieved playability and perfect graphics in-game.

**Comment by Gamemulatorer** [comment]
![Uploading Screenshot_2022-09-01-10-46-22-455_org.ppsspp.ppsspp.jpg…]()
Nice improvement 😁

**Comment by hrydgard** [comment]
Noting to self that when this Vulkan extension is available: https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_attachment_feedback_loop_layout.html  we can just disable all this hackery and do it directly, becuase with that extension (and that image layout) it becomes permitted to render to a different part of the same texture we're reading from, which is exactly what this game does.

**Comment by hrydgard** [comment]
I had another look at this and decided to push it to the next version or something - getting the title screen right and performant will actually be a lot trickier than in-game was. Plenty of higher priority stuff to do.
